docs/zkapps/recursion.mdx

```diff
@@ -0,0 +1,169 @@
+
+## Recursively Verifying a simple program within a zkApp
+
+Let's look at how this works with a simple example that has only one method that proves the public input it received is zero.
+
+```typescript
+import { Field, Experimental } from 'snarkyjs';
+
+let SimpleProgram = Experimental.ZkProgram({
+  publicInput: Field,
+
+  methods: {
+    run: {
+      privateInputs: [],
+
+      method(publicInput: Field) {
+        publicInput.assertEquals(Field.zero);
+      },
+    }
+  }
+});
+```
+
+Next, we compile this program:
+
+```typescript
+let { verificationKey } = await SimpleProgram.compile();
+```
+
+Now we can use it to create a proof:
+
+```typescript
+let proof = await SimpleProgram.run(Field(0));
+```
+
+And we can verify this proof from within any method of your `SmartContract` class:
+
+```typescript
+@method foo(proof: SimpleProgram.Proof) {
+  proof.verify().assertTrue();
+  let output: Field = proof.value;
+  /* ... the rest of our method
+   * For example storing the output of the execution of the program we've
+   * proven inside the public smart contract state. */
+}
+```
+
+What we're doing in `foo` is taking the `SimpleProgram` proof as a private argument to the method, verifying that the execution was valid, and then using the output.
+
+## Recursively Verifying a linear recursive program within a zkApp
+
+This time we are going to write a recursive `ZkProgram` which we can use to create recursive zero-knowledge proofs. In other proof systems, this is extremely difficult to construct if it is even possible, but in SnarkyJS we can describe a recursive ZkProgram with a simple recursive function.
+
+Here we'll write a program that describes a recursive operation of adding one repeatedly to a number. Note that we recursively depend on the older proof as a private argument to our method.
+
+```typescript
+import { SelfProof, Field, Experimental, verify } from 'snarkyjs';
+
+let AddOne = Experimental.ZkProgram({
+  publicInput: Field,
+
+  methods: {
+    baseCase: {
+      privateInputs: [],
+
+      method(publicInput: Field) {
+        publicInput.assertEquals(Field.zero);
+      },
+    },
+
+    step: {
+      privateInputs: [SelfProof],
+
+      method(publicInput: Field, earlierProof: SelfProof<Field>) {
+        earlierProof.verify();
+        earlierProof.publicInput.add(1).assertEquals(publicInput);
+      },
+    },
+  },
+});
+```
+
+First we compile this program and make the base proof as before:
+
+```typescript
+let { verificationKey } = await AddOne.compile();
+
+let proof = await AddOne.baseCase(Field.zero);
+```
+
+This time we can use this proof as input to recursively add one again:
+
+```typescript
+let proof1 = await AddOne.step(Field.one, proof);
+```
+
+And we can repeat this as many times as we want:
+
+```typescript
+let proof2 = await AddOne.step(Field(2), proof1);
+```
+
+Finally we can verify the proof from within a zkApp like we did above:
```

Comment:
* "from within a SmartContract" correct?

Also, this switches from showing how to call ZkProgram, to showing the implementation of the method in the SmartContract. Should probably show how to call it, or provide consistency for the reader.